fuse_core/test/test_throttled_callback.cpp

```diff
@@ -227,35 +255,127 @@ TEST(ThrottledCallback, DropMessagesIfThrottlePeriodIsGreaterThanPublishPeriod)
+  constexpr double kThrottleFrequency = 5.0;
+  constexpr double kThrottlePeriodS = 1 / kThrottleFrequency;
 
-  ASSERT_EQ(nullptr, sensor_model.getLastKeptMessage());
+  const ros::Duration throttled_period(kThrottlePeriodS);
+  PointSensorModel sensor_model(throttled_period, true);
 
-  // Publish some messages:
-  const size_t num_messages = 10;
-  const double period = 0.1 * num_messages / throttled_period.toSec();
-  const double frequency = 1.0 / period;
+  constexpr unsigned kPublishToThrottleRatio = 5.0;
+  constexpr double kPublishFrequency = kThrottleFrequency * kPublishToThrottleRatio;
+  constexpr double kPublishPeriodS = 1 / kPublishFrequency;
 
-  PointPublisher publisher(frequency);
-  publisher.publish(num_messages);
+  PointPublisher publisher(kPublishFrequency);
+
+  constexpr double kBaseTimeS = 10.0;
+  constexpr double kEpsilonS = 1e-3;
+  ros::Time timestamp;
 
-  // Check that regardless of the large throttled period, at least one message is ketpt:
-  EXPECT_EQ(1u, sensor_model.getKeptMessages());
-  EXPECT_EQ(num_messages - 1u, sensor_model.getDroppedMessages());
+  // Initialize the throttled calback with a time just after kBaseTimeS so that kBaseTimeS is the
+  // start timestamp for throttling and the first accepted message will be kThrottlePeriodS after kbaseTimeS
+  timestamp.fromSec(kBaseTimeS + kEpsilonS);
+  publisher.publishWithHeaderTimestamp(timestamp);
+  ros::Duration(0.1).sleep();
 
-  // Check the message kept was the first message:
-  const auto last_kept_message = sensor_model.getLastKeptMessage();
-  ASSERT_NE(nullptr, last_kept_message);
-  EXPECT_EQ(0.0, last_kept_message->x);
+  // First message is dropped, just helps find the next expected message period
+  EXPECT_EQ(sensor_model.getKeptMessages(), 0);
+  EXPECT_EQ(sensor_model.getDroppedMessages(), 1);
+
+  // Make sure that messages before the next start of period are dropped
+  // Publish for a full period starting from just before the start of the previous period
+  // No messages should be kept because we haven't crossed into the period after the initial message yet
+  sensor_model.reset();
+  timestamp.fromSec(kBaseTimeS - kEpsilonS);
+  unsigned i = 0;
+  for (; i < kPublishToThrottleRatio; ++i)
+  {
+    timestamp.fromSec(timestamp.toSec() + kPublishPeriodS);
+    publisher.publishWithHeaderTimestamp(timestamp);
+  }
+
+  ros::Duration(0.1).sleep();
+  EXPECT_EQ(sensor_model.getKeptMessages(), 0);
+  EXPECT_EQ(sensor_model.getDroppedMessages(), i);
+
+  // Publish for another period, phase aligned to just before the period division
+  // Should keep one message
+  sensor_model.reset();
+
+  for (; i < 2 * kPublishToThrottleRatio; ++i)
+  {
+    timestamp.fromSec(timestamp.toSec() + kPublishPeriodS);
+    publisher.publishWithHeaderTimestamp(timestamp);
+  }
+  ros::Duration(0.1).sleep();
+
+  EXPECT_EQ(sensor_model.getKeptMessages(), 1);
+  EXPECT_EQ(sensor_model.getDroppedMessages(), kPublishToThrottleRatio - 1);
```

Comment:
I think it'd be cleaner if you don't use `i` in the `EXPECT_EQ` - you can use `kPublishToThrottleRatio` instead
```suggestion
  sensor_model.reset();

  timestamp.fromSec(kBaseTimeS - kEpsilonS);

  for (std::size_t i = 0; i < kPublishToThrottleRatio; ++i)
  {
    timestamp.fromSec(timestamp.toSec() + kPublishPeriodS);
    publisher.publishWithHeaderTimestamp(timestamp);
  }

  ros::Duration(0.1).sleep();
  EXPECT_EQ(sensor_model.getKeptMessages(), 0);
  EXPECT_EQ(sensor_model.getDroppedMessages(), kPublishToThrottleRatio);

  // Publish for another period, phase aligned to just before the period division
  // Should keep one message
  sensor_model.reset();

  for (std::size_t i = 0; i < kPublishToThrottleRatio; ++i)
  {
    timestamp.fromSec(timestamp.toSec() + kPublishPeriodS);
    publisher.publishWithHeaderTimestamp(timestamp);
  }
  ros::Duration(0.1).sleep();

  EXPECT_EQ(sensor_model.getKeptMessages(), 1);
  EXPECT_EQ(sensor_model.getDroppedMessages(), kPublishToThrottleRatio - 1);
```